datanode/candlesv2/candle_updates.go

```diff
@@ -14,118 +14,122 @@ package candlesv2
 	"sync/atomic"
 	"time"
 
 	"code.vegaprotocol.io/vega/datanode/entities"
+	"code.vegaprotocol.io/vega/datanode/utils"
 	"code.vegaprotocol.io/vega/logging"
 )
 
+var ErrCandleUpdatesStopped = errors.New("candle updates stopped")
+
 type candleSource interface {
 	GetCandleDataForTimeSpan(ctx context.Context, candleID string, from *time.Time, to *time.Time,
 		p entities.CursorPagination) ([]entities.Candle, entities.PageInfo, error)
 }
 
-type subscribeRequest struct {
-	id  string
-	out chan entities.Candle
-}
-
 type CandleUpdates struct {
-	log                *logging.Logger
-	candleSource       candleSource
-	candleID           string
-	subscribeChan      chan subscribeRequest
-	unsubscribeChan    chan string
-	nextSubscriptionID uint64
-	config             CandleUpdatesConfig
+	log          *logging.Logger
+	candleSource candleSource
+	candleID     string
+	config       CandleUpdatesConfig
+	running      atomic.Bool
+	lastCandle   atomic.Pointer[entities.Candle]
+	notifier     utils.Notifier[entities.Candle]
+	stopping     sync.Mutex
 }
 
 func NewCandleUpdates(ctx context.Context, log *logging.Logger, candleID string, candleSource candleSource,
 	config CandleUpdatesConfig) (*CandleUpdates, error,
 ) {
+	observerName := fmt.Sprintf("candles(%s)", candleID)
 	ces := &CandleUpdates{
-		log:             log,
-		candleSource:    candleSource,
-		candleID:        candleID,
-		config:          config,
-		subscribeChan:   make(chan subscribeRequest),
-		unsubscribeChan: make(chan string),
+		log:          log,
+		candleSource: candleSource,
+		candleID:     candleID,
+		config:       config,
+		notifier:     utils.NewNotifier[entities.Candle](observerName, log, 10),
 	}
 
+	ces.running.Store(true)
 	go ces.run(ctx)
 
 	return ces, nil
 }
 
+func (s *CandleUpdates) IsRunning() bool {
+	return s.running.Load()
+}
+
 func (s *CandleUpdates) run(ctx context.Context) {
-	subscriptions := map[string]chan entities.Candle{}
-	defer closeAllSubscriptions(subscriptions)
+	// This is a little bit subtle: We need to ensure that once this goroutine exits, there are
+	// no active subscriptions on our observer, so we take a mutex in s.stop() that prevents
+	// the (unlikely) case that Subscribe() is called while we are shutting down; or else we will
+	// return a channel that will never be serviced.
+	defer s.stop()
 
 	ticker := time.NewTicker(s.config.CandleUpdatesStreamInterval.Duration)
-	var lastCandle *entities.Candle
 
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			if len(subscriptions) > 0 {
-				candles, err := s.getCandleUpdates(ctx, lastCandle)
+			if s.notifier.GetSubscribersCount() > 0 {
+				candles, err := s.getCandleUpdates(ctx, s.lastCandle.Load())
 				if err != nil {
 					s.log.Errorf("failed to get candles, closing stream for candle id %s: %w", s.candleID, err)
 					return
 				}
 
 				if len(candles) > 0 {
-					lastCandle = &candles[len(candles)-1]
+					s.lastCandle.Store(&candles[len(candles)-1])
```

Comment:
Possible race condition here:
Sender Thread-> Store last candle
Subscriber Thread->  Gets last candle sends it  and suscribes
Sender Thread-> Sends all candles including the last candle already sent to all subscribers

I think all you need to do is flip the notify and store last candle order in the sender thread